Introduce option to display shorter author data on EDS search result page

alexklbuckley opened the pull request:
Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley committed:
Introduce option to display shorter author data on EDS search result
page

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley committed:
Display shorter author data on EDS search results: Sync bootstrap3 and
bootstrap5

Sponsored-By: Auckland University of Technology, New Zealand

demiankatz commented on config/vufind/EDS.ini:
```diff
@@ -403,3 +403,14 @@ AU = None
+; display in AuthorNameLimit. If a record has more author names than you define in AuthorNameLimit
+; then VuFind will display the more_authors_abbrev text.
+;[AuthorResultListInfo]
+;ResultListPage = 'Short'
```
If you're going to use mixed-case settings like Short and Long, I'd suggest making the config processing logic case-insensitive to prevent people from getting confused by hard-to-notice configuration subtleties (like s vs. S).

alexklbuckley replied:
Thanks @demiankatz , I've added the config processing case-insensivity you suggested in the next review comment in my latest follow-up. So I think this is now addressed

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +34,34 @@
-                  </h2>
+		  </h2>
+		<?php elseif ('Au' == $item['Group']): ?>
+                    <?php if (isset($this->config->AuthorResultListInfo->ResultListPage) && ($this->config->AuthorResultListInfo->ResultListPage == 'Short')): ?>
```
What if, instead of having a nested if here, you made the check above:

`elseif ('Au' == $item['Group'] && strtolower($this->config->AuthorResultListInfo->ResultListPage ?? 'Long') == 'short'):`

That simplifies the isset, incorporates my case-insensitive suggestion from above, and allows you to utilize the existing else clause below instead of having to make a second nested copy of the logic.

alexklbuckley replied:
thanks for the suggestion and explanation @demiankatz , really appreciate it!

I've included that in my latest follow-up

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +34,34 @@
+                        <?php $authors = $driver->getPrimaryAuthorsWithHighlighting(); ?>
+			<?php $authorCount = count($authors) ?>
+		        <?php $authorNameLimit = $this->config->AuthorResultListInfo->AuthorNameLimit ?? 3; ?>
+                        <?php if (isset($authors) && !empty($authors)): ?>
```
`empty` includes a built-in `isset` check, so you don't need to separately call `isset` here.

alexklbuckley replied:
ahh good to know, thanks Demian. That's included in my latest follow-up

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +34,34 @@
+                        <?php if (isset($authors) && !empty($authors)): ?>
+                            <b><?=$this->transEsc($item['Label'])?>:</b>
+                            <?php $count = 0; ?>
+                            <?php foreach ($authors as $author): ?>
```
Could you use `foreach ($authors as $i => $author)` here and then use `($i + 1)` in place of `$count` to eliminate the need for a separate counter? Or is the author array indexing unreliable/unpredictable?

alexklbuckley replied:
Thanks @demiankatz . I don't beleive author array indexing is unreliable so we should be able to use this. I've added it in my follow-up

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +34,34 @@
+                                    <?=$this->transEsc('more_authors_abbrev')?>
+				    <?php break; ?>
+                                <?php else:?>
+                                    ,
```
Does this syntax create unwanted spaces in front of the commas? I've seen that kind of problem in the past, and it seems like it might be an issue here. If that's the case, you might want to move more of the code into PHP-land and use an echo statement if you need to prevent the code from outputting unwanted whitespace.

alexklbuckley replied:
thanks @demiankatz , yes it does. I've switched to using an echo statement in my follow-up. I wasn't sure if I needed to wrap the comma in a $this->transEsc(','); so I added that just in case.

demiankatz reviewed (changes requested):
Thanks for getting this started, @alexklbuckley! I haven't tried it "hands on" yet, but see below for some suggested optimizations and improvements.

I'd also be interested to hear what @berndoberknapp thinks about this.

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review
requests

- Make config processing logic case-insensitive
- Remove nested if and replace with enhanced elseif
- Remove redundant isset
- Remove separate counter and replace with count in foreach
- Add echo for introducing comma on the end of author names

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review
requests

- Improve usage of the $i for tracking the author index

Sponsored-By: Auckland University of Technology, New Zealand

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +34,30 @@
+                                <?=$this->transEsc('more_authors_abbrev')?>
+			        <?php break; ?>
+                            <?php else:?>
+                                <?php echo $this->transEsc(','); ?>
```
You don't need to `transEsc` the comma -- there is no such translation string. In theory, it might be good to add a translation string named something like `author_list_separator` so this could be more easily translated, but that's probably a separate issue/project. For now I'd just leave this as a plain comma to prevent confusion.

alexklbuckley replied:
thanks @demiankatz I've added that in my latest follow-up!

demiankatz replied:
Do you need to change it back to `<?php echo ','; ?>` to fix the whitespace issues we previously discussed, or is that no longer an issue now?

demiankatz commented on config/vufind/EDS.ini:
```diff
@@ -403,3 +403,14 @@ AU = None
+; If you do set 'ResultListPage' to 'Short' then you can define how many author names to 
+; display in AuthorNameLimit. If a record has more author names than you define in AuthorNameLimit
+; then VuFind will display the more_authors_abbrev text.
+;[AuthorResultListInfo]
```
It might be worth uncommenting the section heading even if leaving the settings commented. It's a common mistake that people will uncomment settings but forget to uncomment sections, so having the section there to begin with can prevent minor annoyance. :-)

alexklbuckley replied:
thanks @demiankatz I've added that in my latest follow-up!

demiankatz reviewed (changes requested):
Thanks -- a couple of minor things. I haven't done hands-on testing of this, but I can do so once these are addressed.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into eds_short_author_data

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review
feedback

- Remove transEsc around comma
- Uncomment new config section heading

Sponsored-By: Auckland University of Technology, New Zealand

demiankatz commented on config/vufind/EDS.ini:
```diff
@@ -403,3 +403,14 @@ AU = None
+; - 'Long' is fetched from the item 'Au' group data.
+; If you do set 'ResultListPage' to 'Short' then you can define how many author names to 
+; display in AuthorNameLimit. If a record has more author names than you define in AuthorNameLimit
+; then VuFind will display the more_authors_abbrev text.
```
We should probably note which option is the default for clarity (though if you might be inclined to change the default behavior, maybe we should wait for that decision to be made before adding the comment -- I'll just leave this note here so we don't forget one way or another).

alexklbuckley replied:
thanks @demiankatz , currently the default behaviour is to continue to display the long author data.

We are happy to continue that so I've added in my latest follow-up that this is the default behaviour. But we are also open to changing that. It would be good to get your thoughts on this  @berndoberknapp?

alexklbuckley replied:
we are happy to proceed with the default being displaying the 'long' author data. As that is in the patchsets I will resolve this thread :)

berndoberknapp replied:
My personal preference would be short as default, but whether we should make such a change probably depends on the target release for this pull request?

I would prefer to also have the option to use short in the detailed record display, which would require a different section name.

alexklbuckley replied:
thanks @berndoberknapp . i've pushed a follow-up setting the default to be 'short' author data for the result list.

Could you , @demiankatz , and @maccabeelevine please retest to ensure that Short is indeed the default on the EDS result list page now?

I will make a separate pull request for adding the option to use short on the detailed record display, under a different section name.

alexklbuckley replied:
perhaps it would be easiest for me to make a new pull request for adding the Short/Long config for the EDS detailed record display once this pull request has been merged. what do you think @demiankatz ?

maccabeelevine replied:
> Could you , @demiankatz , and @maccabeelevine please retest to ensure that Short is indeed the default on the EDS result list page now?

Looks like the default is still 'Long' in the template line 39.

alexklbuckley replied:
my apologies @maccabeelevine . I had not pushed my change to the EDS.ini file, uncommenting the ResultListPage setting and changing it to 'Short'. I've now pushed that change in my latest commit and that resulst-list.phtml template line 39 should now correctly default to 'Short'

maccabeelevine replied:
Hi, I can confirm that the "Short" behavior is now the default if I don't make any changes to EDS.ini.  Thanks!  That said there are two minor issues with the default:

- result-list.phtml line 39 still has 'Long' as the default.  That means if someone doesn't get the EDS.ini update at all, they will still have the Long behavior.  That's common if they have overridden EDS.ini into a local folder and forget to update the overridden version on upgrade.  It's also just inconsistent.
- EDS.ini says `By default, the EDS detailed record page will show the Long author data.`  Which is true in a sense given my previous comment, but I don't think it's what you mean.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +34,30 @@
                     <?=$item['Data']?> </a>
-                  </h2>
+		  </h2>
+		<?php elseif ('Au' == $item['Group'] && strtolower($this->config->AuthorResultListInfo->ResultListPage ?? 'Long') == 'short'): ?> 
```
This doesn't work for me, because `$this->config` is not set in this context. I think you might want to say `$edsConfig = $this->config()->get('EDS');` at the top of the template, and then use `$edsConfig` here....

alexklbuckley replied:
thanks @demiankatz I've added this change to my latest follow-up!

demiankatz reviewed (changes requested):
@alexklbuckley, I did some quick hands-on testing today and ran into a problem (with hopefully an easy solution). See below...

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review
feedback

- Change how EDS config values are fetched by EDS/result-list.phtml.
- Note that the EDS search result list page will show the long author
  data by default.

Sponsored-By: Auckland University of Technology, New Zealand

demiankatz commented:
Thanks for the progress, @alexklbuckley! Now that this is maturing, I'd be interested in hearing whether @maccabeelevine has any thoughts on the current version.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into eds_short_author_data

alexklbuckley committed:
Display shorter author data on EDS search results: Display up to
AuthorNameLimit

- Change a greater than operator in result-list.phtml to greater than or
  equal to. That way if a library sets their EDS.ini AuthorNameLimit = 3
  then only 3 author names will be displayed, not 4.

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley commented:
thanks @demiankatz I noticed one thing while doing local testing. That we were using a greater than operator, rather than greater than or equal to in the template when checking the AuthorNameLimit config.

I've pushed a follow-up  switching to greater than or equal to here: https://github.com/vufind-org/vufind/pull/3841/files#diff-af2922fa62ef85d4a971b9e8fa889f2865dbf710bb535eb719bc7f53f69b8418R49

demiankatz committed:
Minor fixes.

demiankatz commented:
Thanks for the latest fixes, @alexklbuckley. I took another quick look at this and pushed up a couple of small fixes (one to address the unwanted spaces before author separators, and to change the separator from a comma to a semi-colon to improve readability, since the author names often include commas; another to fix the author limit configuration being loaded from the wrong place). I hope this is helpful!

I'll wait to hear opinions from more stakeholders before giving this a closer examination, but I think it's coming along nicely.

alexklbuckley commented:
oh yes, thanks very much for those fixes @demiankatz , appreciate it!

maccabeelevine commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +35,30 @@
+                        <b><?=$this->transEsc($item['Label'])?>:</b>
+                        <?php foreach ($authors as $i => $author): ?>
+                            <a href="<?=$this->record($driver)->getLink('author', $this->highlight($author, null, true, false))?>"><?=$this->highlight($author)?></a><?php
+                            if ($authorCount == ($i + 1)) {
+                              break;
+                            } elseif (($i + 1) >= $authorNameLimit) {
+                              echo ' ' . $this->transEsc('more_authors_abbrev');
+                              break;
+                            } else {
+                              echo '; ';
+                            }
```
You may prefer to extract the ';' logic into CSS (considering it a stylistic element, and perhaps something that would be localized differently), i.e. wrap each `<a>` with some `<span class="author">` and then something like

`.search-results-eds .result .author:not(:last-of-type):after {
    content: '; ';
}`

Then I think the conditional (after the loop) would only have to to deal with adding the more_authors_abbrev if $authorCount > $authorCountLimit.  That said, there's nothing wrong with the way it's implemented now.

alexklbuckley replied:
that's great to hear that you like this feature @maccabeelevine !

I've added this enhancement in my latest follow-up. I am not sure how the CSS works for applying the author name delimiter on the combined result page. So I'll be keen to hear how it looks for you.

maccabeelevine replied:
Yup it works in a combined column!

![image](https://github.com/user-attachments/assets/b40603d9-a56b-4fcf-aea6-c441955d28c3)

alexklbuckley replied:
fabulous, thanks for testing @maccabeelevine !

maccabeelevine commented on config/vufind/EDS.ini:
```diff
@@ -403,3 +403,15 @@ AU = None
 ;url[Expensive Science Database] = https://some.url
 ;url[Cool Humanities Database] = https://another.url
+
+; Define whether to display 'Short' or 'Long' author data on the EDS search result list page.
```
I am seeing the same 'Short' vs 'Long' behavior as you, and I appreciate the detail below about where each kind is extracted from the result.  From looking through the EDS API docs previously, I doubt there is any official explanation of the two data elements.  But if you happen to have found one somewhere, it would be great to link it here for further info.

alexklbuckley replied:
That's great thanks @maccabeelevine 

Unfortunately we haven't found any official docs that we can link. The information in the EDS.ini that we have added is just what we have learnt through debugging the EDS API.

That being the case I will mark this thread as resolved.

maccabeelevine replied:
As expected sadly, but thanks for checking!

alexklbuckley replied:
Not a problem :)

maccabeelevine reviewed (approved):
This is a fantastic feature!  Our institution actually disabled the author in EDS search results because of how unwieldy it can get, even more so in a 'combined' mode (multi-column) layout.  Now we can actually consider restoring it, in short mode.

It seems to work great, just a couple of minor suggestions.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into eds_short_author_data

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review feedback

- Remove adding delimiter between author names in result-list.phtml.
  Extract it into CSS.
- Surround each author name in a <span class="author></span>

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley committed:
Display shorter author data on EDS search results: Add guidance for
setting author name delimiters in CSS

Sponsored-By: Auckland University of Technology, New Zealand

demiankatz commented on config/vufind/EDS.ini:
```diff
@@ -403,3 +403,17 @@ AU = None
+; display in AuthorNameLimit. If a record has more author names than you define in AuthorNameLimit
+; then VuFind will display the more_authors_abbrev text.
+; Also, when you set 'ResultListPage' to 'Short' you can define custom delimiters between the
+; author names in EDS.css file under: 'Author name delimiters'. 
```
Since users may not wish to actually edit/override the full EDS.css file, maybe it would be slightly better to say "by overriding the styles defined in EDS.css" instead of just saying "in EDS.css". Or maybe that's too verbose. Feel free to ignore this suggestion if you don't like it. :-)

alexklbuckley replied:
thanks @demiankatz , I think that is a good addition and I have added that in my latest follow-up :)

berndoberknapp replied:
I would suggest to mention the default, maybe something link "... you can change the default limiter "; " between the author names by overriding...".

alexklbuckley replied:
thanks @berndoberknapp I've added your suggestion in my latest followup.

demiankatz commented:
@alexklbuckley, once you've addressed my latest minor review point, can you take this PR out of draft mode if you're ready for it to be merged?

alexklbuckley committed:
Merge branch 'vufind-org:dev' into eds_short_author_data

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review
feedback

- Clarify in EDS.ini comments how to override author name delimiters.

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley commented:
thanks @demiankatz I've pushed a follow-up addressing your latest minor review point and we are ready for this to be merged so I have taken this PR out of draft mode :)

berndoberknapp commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +35,28 @@
                     <?=$item['Data']?> </a>
-                  </h2>
+		  </h2>
+		<?php elseif ('Au' == $item['Group'] && strtolower($edsConfig->AuthorResultListInfo->ResultListPage ?? 'Long') == 'short'): ?>
```
I would suggest to use 'long' instead of 'Long' to match the strtolower and the 'short'. Same for bootstrap5.

alexklbuckley replied:
thanks @berndoberknapp I have prioritised your note about making 'Short' the default option for the Result list page, as you will see in my code now. So this should no longer be needed.

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review
feedback

- Clarify in the EDS.ini comments that the default delimiter between author names is "; "

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review
feedback

- Switch default EDS result list author name display to 'Short'

Sponsored-By: Auckland University of Technology, New Zealand

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +35,28 @@
+		    <?php $authorCount = count($authors) ?>
+		    <?php $authorNameLimit = $edsConfig->AuthorResultListInfo->AuthorNameLimit ?? 3; ?>
+                    <?php if (!empty($authors)): ?>
+                        <b><?=$this->transEsc($item['Label'])?>:</b>
```
I notice that the else clause is inside a `<p>` tag but this case is not. Is that intentional, or should the `<p>` be added to match?

alexklbuckley replied:
thanks @demiankatz ! Ive included that in my latest follow up

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +35,28 @@
+                    <?php $authors = $driver->getPrimaryAuthorsWithHighlighting(); ?>
+		    <?php $authorCount = count($authors) ?>
+		    <?php $authorNameLimit = $edsConfig->AuthorResultListInfo->AuthorNameLimit ?? 3; ?>
+                    <?php if (!empty($authors)): ?>
```
Since you've already done the work to count the authors, I wonder if it's more efficient to use `$authorCount > 1` here instead of doing an `empty($authors)` check, since `empty` is likely more work.

alexklbuckley replied:
thanks @demiankatz ! Ive included that in my latest follow up

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley -- see below for a couple more small suggestions.

If it's easier to create a separate PR to address the record view, that's fine with me. I'm happy to merge this as an incremental step once everyone agrees that it's ready. We should just think ahead of time about whether it makes sense to have separate settings for each location, or a single shared setting. I think flexibility is a good thing, but it would be nice to have an easy way to change both -- e.g. if one setting is set and the other is commented out, maybe they could default to each other, if that makes any sense. Whatever we do, we should document as clearly as possible. :-)

alexklbuckley committed:
Merge branch 'vufind-org:dev' into eds_short_author_data

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review feedback

- Switch conditional from empty($authors) to $authorCount > 1
- Surround else clause with <p> and </p> tags

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review feedback

- Switch to a default short author data display on EDS result list

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review feedback

- Try fix indentation on template.

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley commented:
> Thanks, @alexklbuckley -- see below for a couple more small suggestions.
> 
> If it's easier to create a separate PR to address the record view, that's fine with me. I'm happy to merge this as an incremental step once everyone agrees that it's ready. We should just think ahead of time about whether it makes sense to have separate settings for each location, or a single shared setting. I think flexibility is a good thing, but it would be nice to have an easy way to change both -- e.g. if one setting is set and the other is commented out, maybe they could default to each other, if that makes any sense. Whatever we do, we should document as clearly as possible. :-)

Thanks @demiankatz ! 

Yes, from our perspective splitting this over two pull requests would be helpful and appreciated. 

Yes, that is a very good point about if there should be one shared setting or a separate setting for each location. 

I agree that having flexibility is good. I think your suggestion of if one setting is set, and the other is commented out, then VuFind only checks the uncommented value is a good one. I would happy with that. 

If you could please let me know if anymore changes are required on this pull request that would be great. 

Thanks as always!

maccabeelevine commented on themes/bootstrap3/templates/RecordDriver/EDS/result-list.phtml:
```diff
@@ -34,7 +35,30 @@
                     <a href="<?=$this->escapeHtmlAttr($recordLinker->getUrl($this->driver))?>" class="title getFull _record_link"  data-view="<?=$this->escapeHtmlAttr($this->params->getOptions()->getListViewOption())?>">
                     <?=$item['Data']?> </a>
-                  </h2>
+		  </h2>
```
I'm seeing a mix of tabs and spaces for indentation for a few lines starting here.  I think spaces are the standard for VuFind, but @demiankatz shouldn't php-cs-fixer or one of the other checks be catching this in the automated tests?

demiankatz replied:
@maccabeelevine, I think what's happening is that we're relying on `phpcbf` rather than `php-cs-fixer` for indentation normalization, but we don't apply `phpcbf` to templates. It looks like we've [explicitly disabled the statement_identation php-cs-fixer rule for templates](https://github.com/vufind-org/vufind/blob/dev/tests/vufind_templates.php-cs-fixer.php#L63) because it was misbehaving and garbling the files. This may be causing us to miss out on some issues. I'm not sure if anything in [this StackOverflow thread](https://stackoverflow.com/questions/42849484/use-tab-for-indentation-in-php-cs-fixer) might be used to improve the whitespace formatting without breaking the overall file indentation; it might be worth experimenting with. I don't have time at this very moment, but if nobody else is interested in digging in deeper, let me know and I can put it on my to-do list for later.

maccabeelevine replied:
Thanks @demiankatz.  I followed that thread's advice (and some others) but no luck -- I can get it to change n!=2 spaces to 2, but not to change tabs :)  At least not in HTML.  Not any priority.

demiankatz replied:
Seems very possible that php-cs-fixer intentionally ignores anything outside of `<?php` tags, in which case we would need a different tool to solve HTML formatting issues in our templates.

alexklbuckley replied:
Thanks @demiankatz and @maccabeelevine , I've pushed fixes for those two points in https://github.com/vufind-org/vufind/pull/3841#discussion_r1713666675 . If you could please re-review now?

maccabeelevine reviewed (changes requested):
Just minor stuff

demiankatz commented:
Thanks, @alexklbuckley -- I see that @maccabeelevine has just left some feedback, so I'll wait until you address his comments before I take another look at this, so I can approach the code with fresher eyes. :-)

alexklbuckley committed:
Merge branch 'vufind-org:dev' into eds_short_author_data

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review feedback

- Fix wording about default option in EDS.ini
- Change default in result-list.phtml line 39

Sponsored-By: Auckland University of Technology, New Zealand

maccabeelevine reviewed (approved):
Everything I suggested looks good now!  Looking forward to integrating this new feature in our instance.

demiankatz commented on config/vufind/EDS.ini:
```diff
@@ -403,3 +403,18 @@ AU = None
 ;url[Expensive Science Database] = https://some.url
 ;url[Cool Humanities Database] = https://another.url
+
+; Define whether to display 'Short' or 'Long' author data on the EDS search result list page.
```
I'm glad that the functionality is looking good now (thanks for re-testing, @maccabeelevine!). Before we merge, though, I think it might make sense to adjust the configuration a bit to make it more future-proof for the intended next step. I'd suggest that perhaps we could revise the configs to something like this:

```
; This section controls how the authors of EDS records are displayed.
[AuthorDisplay]
; Control which data is used to display author information in search results.
; - 'Short' (the default) is fetched from the RecordInfo section of a record (the 
;   'BibRecord/BibRelationships/HasContributorRelationships/*PersonEntity/Name/NameFull' data).
;   Also, when you set 'ResultListPage' to 'Short' you can change the default delimiter "; " between
;   the author names by overriding the styles defined in the EDS.css file
;   under: 'Author name delimiters'.
; - 'Long' is fetched from the item 'Au' group data.
ResultListFormat = 'Short'
; This setting controls how many author names to display when ResultListFormat is
; set to 'Short'. It is ignored if the setting is 'Long'. If a record has more author names
; than you define in AuthorNameLimit, then VuFind will display the more_authors_abbrev text.
;AuthorNameLimit = 3
```
...and then also of course adjust the code to use the new config setting names.

Note that this is just an initial proposal, and there might be other/better ways to adjust the config. The idea here is that we have location-specific "format" settings (so ResultListFormat now, and in the future, RecordFormat as well -- and maybe a DefaultFormat as fallback) but we have only one AuthorNameLimit. The AuthorNameLimit is applied to "Short" but not "Long" which makes logical sense (the "Short" list is truncated). But that raises a new question: do we need separate settings for data source vs. length? i.e. might you want a "Short" (truncated) list that's sourced from the "Long" ('Au' group) data? I'm not really sure -- but if that kind of scenario is conceivable, maybe we really want separate DataSource and Format settings, to separate the data source from the display format. Or maybe we want the AuthorNameLimit to be differentiated by location in parallel with the Format setting, instead of being tied to "Short" format.

What do you think? I don't think it will be too hard to implement any of these proposed ideas, but I think it's a good idea to agree on the most useful/logical configuration format before we merge this, so the follow-up work will be simple and straightforward to implement, and we won't end up changing things that have already been committed.

berndoberknapp replied:
The format of the authors in Au can differ significantly depending on the content provider, therefore I don't think we should try to apply the limit in that case. Maybe ShortAuthorLimit instead of AuthorNameLimit?

alexklbuckley replied:
thanks @demiankatz and @berndoberknapp !

Sorry for my delay providing a followup.

What I've added in my latest follow-up is:

1. Added section name of [AuthorDisplay] as you suggested @demiankatz  - and made the code use it.
2. Added config name ResultListFormat as you suggested @demiankatz - and made the code use it.
3. Added the config name ShortAuthorLimit, as you suggested @berndoberknapp  - and made the code use it.

--
I think what we could do in the follow-up commit is:
- Add a DetailPageFormat config if that is also set to 'Short' then VuFind looks at the value of  the single ShortAuthorLimit config to decide how many author names to display.

i.e. we could have a config like so:
```
[AuthorDisplay] 
ResultListFormat = 'Short'
DetailPageFormat = 'Short'
ShortAuthorLimit = 3
```

Could you please let me know your thoughts on this idea?

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley! In looking at this again with fresh eyes, I think we should clarify the configurations a bit here before we take the next step of merging this and adding more settings. See below for more detailed ideas.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into eds_short_author_data

alexklbuckley committed:
Merge branch 'vufind-org:dev' into eds_short_author_data

alexklbuckley committed:
Display shorter author data on EDS search results: Implement review feedback

- Change EDS config section name to [AuthorDisplay]
- Change EDS config setting name to ResultListFormat
- Change EDS config setting name to ShortAuthorLimit

Sponsored-By: Auckland University of Technology, New Zealand

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into eds_short_author_data

demiankatz committed:
Add default value comment.

demiankatz committed:
Simplify code; fix whitespace.

demiankatz reviewed (approved):
Thanks, @alexklbuckley, this seems to work as expected for me, and I think your adjustments to the configuration are reasonable. I've made a couple of small adjustments while testing:

1.) I updated EDS.ini to note the default value of ShortAuthorLimit, since that was not previously documented.

2.) I did some refactoring and reformatting of result-list.phtml; some of the whitespace was a mix of tabs and spaces which made it hard to read, so I converted it to all-spaces. Additionally, I slightly refactored the if/else related to the author count to eliminate redundant code opening and closing the paragraph tab and displaying the label.

I'll hold off on merging this until you've had a chance to review my changes, and in case @berndoberknapp or @maccabeelevine have any further suggestions. Please let me know when you think this is fully ready for merge.

alexklbuckley commented:
> Thanks, @alexklbuckley, this seems to work as expected for me, and I think your adjustments to the configuration are reasonable. I've made a couple of small adjustments while testing:
> 
> 1.) I updated EDS.ini to note the default value of ShortAuthorLimit, since that was not previously documented.
> 
> 2.) I did some refactoring and reformatting of result-list.phtml; some of the whitespace was a mix of tabs and spaces which made it hard to read, so I converted it to all-spaces. Additionally, I slightly refactored the if/else related to the author count to eliminate redundant code opening and closing the paragraph tab and displaying the label.
> 
> I'll hold off on merging this until you've had a chance to review my changes, and in case @berndoberknapp or @maccabeelevine have any further suggestions. Please let me know when you think this is fully ready for merge.

Thanks for your feedback and implementing those changes @demiankatz .

I've had a look over them and they look good to me!

berndoberknapp reviewed (approved):
Looks good to me, too!

demiankatz commented:
Thanks, everyone! Since we now have approvals across the board, I'll go ahead and merge this. Any further feedback can be incorporated into the follow-up PR we discussed to add more settings for the record page.

alexklbuckley commented:
Thanks very much @demiankatz , @maccabeelevine and @berndoberknapp !

alexklbuckley commented:
Hi all, noting that I am working on the follow-up PR we discussed to add more settings to the record page: https://github.com/vufind-org/vufind/pull/3881
